runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/Semaphore.java

```diff
@@ -76,6 +82,11 @@ public int hashCode() {
 	}
 
 	public synchronized void release() {
+		if (notifications > 0) {
+			System.err.println("Semaphore already released"); //$NON-NLS-1$
```

Comment:
Actually a Semaphore holds a positive number of permits and thus can even has notifications = 5, as there is no way for the code to init the initial number of permits it seems that the intend is here to

1. construct the semaphore
2. call `release` the number of permits one likes to allow (e.g. 5 times)
3. now there should always be symmetric calls to aquire/release

Reply:
thanks for the explanation, would be good if some javadoc explains something.

i just had a deadlocked VM and a thread was stuck in an endless wait in acquire with notifications=0 and i know nothing  how this is meant to work. so i just wanted to know what the CI says.

Reply:
I don't know *why* this class is there and *where* it is used, so I can only guess from the name, but actually java has also a sempahore class (it says since 1.5): https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html that has a rather good description about Sempahores...

So usually an endless wait on `notifications=0` means no one likes to release the permit (or the semaphore never has had any what would be quite strange).

Reply:
There is only two usages:
# OrderedLock.operations- as a Lock - like java.util.concurrent.locks.ReentrantLock(true)?
# JobManager - as a barrier  - like java.util.concurrent.CyclicBarrier ?

Back when eclipse was written those java classes may have not been existing in the jdk yet. But since there are standard classes now, i would favor the use of those.

Reply:
Sure standard java classes are often more complete, better tested and better integrate with other parts of the system and are likely better understood by other developers.

So as this is an internal class, as soon as all references are replaced, the class can be deleted.